File: samples/FSharp/Database.Mapped.fs
namespace Samples.FSharp

open System
open System.Collections
open System.Collections.Generic
open System.Data.Linq.Mapping
open System.IO
open System.Transactions
open DbExtensions
open Samples.FSharp.Northwind

type DatabaseMappedSamples(connectionString : string, mapping : MetaModel, log : TextWriter) =

   let db = 
      let db1 = new NorthwindDatabase(connectionString, mapping) 
      db1.Configuration.Log <- log
      db1

   member this.IncludeManyToOne() =
         
      db.Products
         .Include("Category")
         .Include("Supplier")
         .Take(3)
         .AsEnumerable()

   member this.IncludeManyToOneNested() =

      db.EmployeeTerritories
         .Include("Territory.Region")
         .Take(3)
         .AsEnumerable()

   member this.IncludeOneToMany() =

      db.Regions
         .Include("Territories")
         .First()

   member this.ContainsKey() =
      db.Products.ContainsKey(1)

   member this.Find() =
      db.Products.Find(1)

   member this.Transactions_AdoNet() =

      using (db.EnsureInTransaction())(fun tx ->
         // Connection is automatically opened if not open

         this.Transactions_DoWork()

         tx.Commit()
      )
      // Connection is closed if wasn't open

   member this.Transactions_TransactionScope() =

      using (new TransactionScope())(fun tx -> 
         using (db.EnsureConnectionOpen())(fun x ->
            // Open connection if not open

            this.Transactions_DoWork()

            tx.Complete()
         )
         // Connection is closed if wasn't open
      )

   member private this.Transactions_DoWork() =

      let order = new Order(CustomerID = "ALFKI")
      order.OrderDetails.Add(new OrderDetail(ProductID = 77, Quantity = 1s))
      order.OrderDetails.Add(new OrderDetail(ProductID = 41, Quantity = 2s))

      db.Orders.Insert(order, deep = true)

      order.Freight <- new Nullable<decimal>(10m)

      db.Orders.Update(order)

      // The following line is not needed when cascade delete is configured on the database
      db.OrderDetails.DeleteRange(order.OrderDetails);

      db.Orders.Delete(order)

File: samples/CSharp/Database.Mapped.cs
using System;
using System.Collections.Generic;
using System.Data.Linq.Mapping;
using System.Diagnostics;
using System.IO;
using System.Transactions;
using Samples.CSharp.Northwind;

namespace Samples.CSharp {

   public class DatabaseMappedSamples {

      readonly NorthwindDatabase db;

      public DatabaseMappedSamples(string connectionString, MetaModel mapping, TextWriter log) {

         this.db = new NorthwindDatabase(connectionString, mapping) {
            Configuration = { 
               Log = log
            }
         };
      }

      public IEnumerable<Product> PredicateOnly() {
         return db.Products.Where("ProductID <= {0}", 7).AsEnumerable();
      }

      public bool ContainsKey() {
         return db.Products.ContainsKey(1);
      }

      public Product Find() {
         return db.Products.Find(1);
      }

      public void Refresh() {

         Product product = db.Products.Find(1);

         Debug.Assert(product.ProductName == "Chai");

         product.ProductName = "xxx";

         db.Products.Refresh(product);

         Debug.Assert(product.ProductName == "Chai");
      }
      
      public void Transactions_AdoNet() {

         using (var tx = db.EnsureInTransaction()) {
            // Connection is automatically opened if not open

            Transactions_DoWork();

            tx.Commit();
         }
         // Connection is closed if wasn't open
      }

      public void Transactions_TransactionScope() {

         using (var tx = new TransactionScope()) {
            using (db.EnsureConnectionOpen()) {
               // Open connection if not open

               Transactions_DoWork();

               tx.Complete();
            }
            // Connection is closed if wasn't open
         }
      }

      void Transactions_DoWork() {

         var order = new Order {
            CustomerID = "ALFKI",
            OrderDetails = { 
               new OrderDetail { ProductID = 77, Quantity = 1 },
               new OrderDetail { ProductID = 41, Quantity = 2 }
            }
         };

         db.Orders.Insert(order, deep: true);

         order.Freight = 10m;

         db.Orders.Update(order);

         // The following line is not needed when cascade delete is configured on the database
         db.OrderDetails.DeleteRange(order.OrderDetails);

         db.Orders.Delete(order);
      }
   }
}